go/libkb/merkle_client.go

```diff
@@ -706,54 +775,66 @@ func (mc *MerkleClient) lookupRootAndSkipSequence(m MetaContext, lastRoot *Merkl
 	return mr, ss, apiRes, err
 }
 
-func (mc *MerkleClient) lookupPathAndSkipSequenceUser(m MetaContext, q HTTPArgs, sigHints *SigHints, lastRoot *MerkleRoot, opts MerkleOpts) (vp *VerificationPath, ss SkipSequence, userInfo *merkleUserInfoT, apiRes *APIRes, err error) {
+func (mc *MerkleClient) lookupLeafAndPathUser(m MetaContext, q HTTPArgs, sigHints *SigHints, root *MerkleRoot, opts MerkleOpts) (vp *VerificationPath, userInfo *merkleUserInfoT, apiRes *APIRes, err error) {
 	opts.isUser = true
-	apiRes, err = mc.lookupPathAndSkipSequenceHelper(m, q, sigHints, lastRoot, opts)
+	vp, apiRes, err = mc.lookupLeafAndPath(m, q, root, sigHints, opts)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	if sigHints != nil {
 		if err = sigHints.RefreshWith(m, apiRes.Body.AtKey("sigs")); err != nil {
-			return nil, nil, nil, nil, err
+			return nil, nil, nil, err
 		}
 	}
 
-	vp, userInfo, err = mc.readPathFromAPIResUser(m, apiRes)
-	if err != nil {
-		return nil, nil, nil, nil, err
-	}
-
-	ss, err = mc.readSkipSequenceFromAPIRes(m, apiRes, vp.root, lastRoot)
+	userInfo, err = mc.readUserFromAPIRes(m, apiRes)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return nil, nil, nil, err
 	}
 
-	return vp, ss, userInfo, apiRes, nil
+	return vp, userInfo, apiRes, nil
 }
 
-func (mc *MerkleClient) lookupPathAndSkipSequenceTeam(m MetaContext, q HTTPArgs, lastRoot *MerkleRoot, opts MerkleOpts) (vp *VerificationPath, ss SkipSequence, res *APIRes, err error) {
-	apiRes, err := mc.lookupPathAndSkipSequenceHelper(m, q, nil, lastRoot, opts)
+func (mc *MerkleClient) lookupLeafAndPath(m MetaContext, q HTTPArgs, root *MerkleRoot, sigHints *SigHints, opts MerkleOpts) (vp *VerificationPath, res *APIRes, err error) {
+	apiRes, root, err := mc.lookupLeafAndPathHelper(m, q, sigHints, root, opts)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, err
 	}
 
 	vp, err = mc.readPathFromAPIRes(m, apiRes, opts)
 	if err != nil {
-		return nil, nil, nil, err
+		return nil, nil, err
 	}
+	vp.root = root
 
-	ss, err = mc.readSkipSequenceFromAPIRes(m, apiRes, vp.root, lastRoot)
-	if err != nil {
-		return nil, nil, nil, err
+	return vp, apiRes, nil
+}
+
+// `isUser` is true for loading a user and false for loading a team.
+func (mc *MerkleClient) lookupLeafAndPathHelper(m MetaContext, q HTTPArgs, sigHints *SigHints, root *MerkleRoot, opts MerkleOpts) (apiRes *APIRes, newRoot *MerkleRoot, err error) {
+	defer m.VTrace(VLog1, "MerkleClient#lookupLeafAndPathHelper", func() error { return err })()
+
+	apiRes, rootRefreshNeeded, err := mc.lookupLeafAndPathHelperOnce(m, q, sigHints, root, opts)
+
+	for rootRefreshNeeded {
```

Comment:
As @heronhaye suggested, I think it's worth to break out of here after ~5 tries, and error out.